### PR TITLE
Fud interpreter integration

### LINKED ISSUE #560: Allow interpreter to consume .data (json) files
We should be able to initialize memory state from data files to allow evaluation of programs relying on input & output memories.

### PR DESCRIPTION
**Pull request by EclecticGriffin**
An attempt to close #560.

Major changes include:

***fud support for invoking the interpreter***
  try `fud e --to interpreter-out ...`

***fud support for passing data files to the interpreter***
  try `fud exec --to interpreter-out tests/while.futil -s interpreter.data tests/while.futil.data `

***Interpreter can now take and initialize memories based on the data marshaled from fud***

***memories are now a flat vec with row-major ordering indexing***

***NEW SERIALIZATION CRIMES***

Anyway, this still needs a bit of work, mainly in the writing tests department. But also someone please check my work wrt the row-major ordering calculations because I don't know that I did them correctly.

Also there's a bug in `tests/while.futil` for the interpreter since the addr0 of the memory is never set and the cond group doesn't assign to it while reading. The solution would be to give the initial value of a memory's read_data port the value of the zero index, or we could just decide this program is bad and should feel bad?

Anyway, my brain feels like spaghetti, so I'm going to stop writing words.

### CODE REVIEW

**Comment by EclecticGriffin** [comment]
Also a little demo:

```
$ fud exec --to interpreter-out tests/while.futil -s interpreter.data tests/while.futil.data                                                                                                                               
{
  "ports": {
    "main": {
      "_0_1": {
        "out": 0
      },
      "_1_1": {
        "out": 1
      },
      "_1_32": {
        "out": 1
      },
      "_8_32": {
        "out": 8
      },
      "add": {
        "left": 0,
        "out": 0,
        "right": 0
      },
      "i": {
        "addr0": 0,
        "clk": 0,
        "done": 0,
        "read_data": 148,
        "write_data": 0,
        "write_en": 0
      },
      "lt": {
        "left": 0,
        "out": 0,
        "right": 0
      }
    }
  },
  "memories": {
    "main": {
      "i": [
        148
      ]
    }
  }
}
```

File: interp/src/main.rs [hunk 1]
```diff
@@ -26,6 +26,10 @@ pub struct Opts {
     /// Path to the primitives library
     #[structopt(long, short, default_value = "..")]
     pub lib_path: PathBuf,
+
+    /// Path to optional datafile
```
Comment (rachitnigam) [inline interp/src/main.rs:30]: Comment should say what happens if the optional is not set.

File: fud/fud/stages/interpreter.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+                if round_float_to_fixed:
+                    # Only round if it is not already representable.
+                    fractional_width = width - int_width
+                    x = float_to_fixed_point(float(x), fractional_width)
+                    x = str(x)
+                    return FixedPoint(x, **shape[k]).bit_string(with_prefix)
+                else:
+                    raise error
+
+        output_json[k] = list([convert(x) for x in arr.flatten()])
```
Comment (sampsyo) [inline fud/fud/stages/interpreter.py:142]: FWIW, I don't think the `list(...)` call is necessary (the `[]` comprehension already produces a list).

File: interp/src/primitives.rs [hunk 3]
```diff
@@ -672,6 +680,20 @@ impl StdMemD2 {
...
+
+        for (idx, val) in vals.iter().enumerate() {
+            assert_eq!(val.len(), self.width as usize);
+            self.data[idx] = val.clone()
+        }
+    }
+
+    #[inline]
+    fn calc_addr(&self, addr0: u64, addr1: u64) -> u64 {
+        addr1 + addr0 * self.d1_size
```
Comment (sampsyo) [inline interp/src/primitives.rs:695]: LGTM! Super trivial comment here, and feel free to ignore, but I usually see this kind of calculation written the other way: `addr0 * self.d1_size + addr1`. It sort of emphasizes that, as you walk down the array, the rightmost thing changes the fastest.

File: interp/src/primitives.rs [hunk 4]
```diff
@@ -869,6 +893,23 @@ impl StdMemD3 {
...
+
+        for (idx, val) in vals.iter().enumerate() {
+            assert_eq!(val.len(), self.width as usize);
+            self.data[idx] = val.clone()
+        }
+    }
+
+    #[inline]
+    fn calc_addr(&self, addr0: u64, addr1: u64, addr2: u64) -> u64 {
+        addr2 + self.d2_size * (addr1 + addr0 * self.d1_size)
```
Comment (sampsyo) [inline interp/src/primitives.rs:911]: Also LGTM! This is harder to reverse, of course, but if you want:

`addr0 * self.d1_size * self.d2_size + addr1 * self.d2_size + addr2`

…sorta has a rhythm to it.

File: interp/src/primitives.rs [hunk 5]
```diff
@@ -1093,6 +1137,26 @@ impl StdMemD4 {
...
+            assert_eq!(val.len(), self.width as usize);
+            self.data[idx] = val.clone()
+        }
+    }
+
+    #[inline]
+    fn calc_addr(&self, addr0: u64, addr1: u64, addr2: u64, addr3: u64) -> u64 {
+        addr3
+            + self.d3_size
+                * (addr2 + self.d2_size * (addr1 + addr0 * self.d1_size))
```
Comment (sampsyo) [inline interp/src/primitives.rs:1158]: Also great, AFAICT!

**Review by sampsyo — COMMENTED** [review]
Super awesome!! Seems like all the index math is right from here. :tada:

File: fud/fud/config.py [hunk 6]
```diff
@@ -26,6 +26,12 @@
             "file_extensions": [".futil"],
             "flags": None,
         },
+        "interpreter": {
+            "exec": "./target/debug/interp",
+            "flags": None,
+            "data": None,
+            "round_float_to_fixed": True,
+        },
```
Comment (rachitnigam) [inline fud/fud/config.py:34]: Add a new check for `fud check` to warn users if this has not been built.

File: fud/fud/stages/interpreter.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import numpy as np
+from fud.stages.verilator.numeric_types import FixedPoint, Bitnum
+from fud.errors import InvalidNumericType
+from fud.stages.verilator.json_to_dat import parse_fp_widths
+from calyx.utils import float_to_fixed_point
+
+
+from ..utils import shell, TmpDir
+
+_FILE_NAME = "data.json"
```
Comment (rachitnigam) [inline fud/fud/stages/interpreter.py:13]: This global constant is used. Why is it prefixed with `_`?

**Review by rachitnigam — COMMENTED** [review]
Few comments for fud stuff.

One broader comment: we should have consistent behavior w.r.t to the `@external` attribute---memories that are not marked with `@external` cannot be initialized by the `data.json` file and are not printed out when evaluation completes. Does not need to be in this PR.

File: fud/fud/stages/interpreter.py (hunk 7 above)
Comment (EclecticGriffin) [reply fud/fud/stages/interpreter.py:13]: It's just to indicate a "private" variable not needed outside this file. It's been a while since I've written python, but I believe this is the accepted syntax?

**Comment by EclecticGriffin** [comment]
Could you elaborate on the `@external` stuff in an issue?

File: fud/fud/stages/interpreter.py (hunk 2 above)
Comment (EclecticGriffin) [reply fud/fud/stages/interpreter.py:142]: Lol, how my python has atrophied

**Comment by rachitnigam** [comment]
I can attempt to explain but try the [documentation first](https://capra.cs.cornell.edu/docs/calyx/lang/attributes.html#external1).

TLDR: It is a mechanism to specify which memories to print out and initialize. Every other memory is not printed out.

File: fud/fud/stages/interpreter.py (hunk 7 above)
Comment (rachitnigam) [reply fud/fud/stages/interpreter.py:13]: Got it. A comment above the constant to saying this would be useful?

Comment (EclecticGriffin) [reply fud/fud/stages/interpreter.py:15]: 👍🏻

**Comment by EclecticGriffin** [comment]
Okay so basically, we only want to print out the memory state of memory cells marked with `@external` (and all registers?) ?

File: fud/fud/config.py (hunk 6 above)
Comment (EclecticGriffin) [reply fud/fud/config.py:34]: Fud check already prints a warning if the interpreter path doesn't point to anything. Were you looking for something else?

**Comment by EclecticGriffin** [comment]
Anyway, this should be good to go, pending the resolution of the `fud check` conversation

File: fud/fud/config.py (hunk 6 above)
Comment (sampsyo) [reply fud/fud/config.py:34]: I think that's probably all that's necessary!

Comment (rachitnigam) [reply fud/fud/config.py:34]: Oh, you’re right. This sounds good.

**Comment by rachitnigam** [comment]
Yes to the understanding of @external

**Comment by EclecticGriffin** [comment]
Okay great, I can open an issue for that then since we'll probably want to address that in a different PR. But otherwise sounds like this PR is good to go?
